server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1640,6 +1656,72 @@ public void snapshotShard(Store store, MapperService mapperService, SnapshotId s
+     * alongside their content as hash in the {@link BlobStoreIndexShardSnapshots} metadata to quickly load all {@link SegmentInfos} for
+     * each shard snapshot.
+     * Also see {@link StoreFileMetaData#hashEqualsContents()}.
+     *
+     * @param snapshotIndexCommit Index commit to snapshot
+     * @param snapshots shard snapshots already in the repository
+     * @return List of files already in the repository to use as the given shard's snapshot or {@code null} if no such set of files could
+     *         be found in the current shard snapshots
+     */
+    @Nullable
+    private static List<BlobStoreIndexShardSnapshot.FileInfo> findMatchingShardSnapshot(
+            IndexCommit snapshotIndexCommit, BlobStoreIndexShardSnapshots snapshots) {
+        final Map<String, String> userCommitData;
+        try {
+            userCommitData = snapshotIndexCommit.getUserData();
+        } catch (IOException e) {
+            assert false : new AssertionError("Did not expect #getUserData to throw but saw exception", e);
+            return null;
+        }
+        final String sequenceNum = userCommitData.get(SequenceNumbers.MAX_SEQ_NO);
+        final String localCheckpoint = userCommitData.get(SequenceNumbers.LOCAL_CHECKPOINT_KEY);
+        final String primaryTerm = userCommitData.get(Engine.MAX_PRIMARY_TERM);
+        final String historyUUID = userCommitData.get(Engine.HISTORY_UUID_KEY);
+        if (sequenceNum != null && localCheckpoint != null && primaryTerm != null && historyUUID != null) {
+            for (SnapshotFiles snapshotFileSet : snapshots.snapshots()) {
+                final List<BlobStoreIndexShardSnapshot.FileInfo> files = snapshotFileSet.indexFiles();
+                final SegmentInfos segmentInfos;
+                try {
+                    final Directory dir = new ByteBuffersDirectory();
+                    for (BlobStoreIndexShardSnapshot.FileInfo f : files) {
+                        final StoreFileMetaData m = f.metadata();
+                        if (m.hashEqualsContents() == false) {
+                            continue;
+                        }
+                        try (IndexOutput indexOutput = dir.createOutput(m.name(), IOContext.DEFAULT)) {
+                            final BytesRef fileContent = m.hash();
+                            indexOutput.writeBytes(fileContent.bytes, fileContent.offset, fileContent.length);
+                        }
+                    }
+                    segmentInfos = SegmentInfos.readLatestCommit(dir);
+                } catch (IOException e) {
+                    logger.debug("Failed to read SegmentInfos from files {}", files);
+                    continue;
+                }
+                final Map<String, String> snapshotUserCommitData = segmentInfos.getUserData();
+                if (sequenceNum.equals(snapshotUserCommitData.get(SequenceNumbers.MAX_SEQ_NO)) &&
```

Comment:
Should we add maxPrimaryTerm to SequenceNumbers#CommitInfo then use its equals here?